simulator/source/Combat_simulator.cpp

```diff
@@ -1169,6 +1168,29 @@ void Combat_simulator::swing_weapon(Weapon_sim& weapon, Weapon_sim& main_hand_we
     }
 }
 
+void update_swing_timers(Weapon_sim& mh, Weapon_sim& oh, double oldHaste, double haste)
+{
+    if (mh.internal_swing_timer <= 0)
+    {
+        mh.internal_swing_timer = mh.swing_speed / (1 + haste);
+    }
+    else if (haste != oldHaste)
+    {
+        mh.internal_swing_timer *= (1 + oldHaste) / (1 + haste);
+    }
+
+    if (oh.socket != Socket::off_hand) return;
+
+    if (oh.internal_swing_timer <= 0)
+    {
+        oh.internal_swing_timer = oh.swing_speed / (1 + haste);
+    }
+    else if (haste != oldHaste)
+    {
+        oh.internal_swing_timer *= (1 + oldHaste) / (1 + haste);
+    }
+}
```

Comment:
this is simply factored out, so it can be called from within the "use_slam" block, too